python/cuml/test/test_pickle.py

```diff
@@ -0,0 +1,283 @@
+# limitations under the License.
+#
+
+import pytest
+import os
+import cuml
+from cuml.test.utils import array_equal, np_to_cudf
+import cudf
+import numpy as np
+import pandas as pd
+from sklearn.datasets import make_regression, make_blobs
+import pickle
+import dill
+
+regression_models = dict(
+    LinearRegression=cuml.LinearRegression(),
+    Lasso=cuml.Lasso(),
+    Ridge=cuml.Ridge(),
+    ElasticNet=cuml.ElasticNet()
+)
+
+solver_models = dict(
+    CD=cuml.CD(),
+    SGD=cuml.SGD(eta0=0.005)
+)
+
+cluster_models = dict(
+    KMeans=cuml.KMeans()
+)
+
+decomposition_models = dict(
+    PCA=cuml.PCA(),
+    TruncatedSVD=cuml.TruncatedSVD(),
+    UMAP=cuml.UMAP()
+)
+
+neighbor_models = dict(
+    NearestNeighbors=cuml.NearestNeighbors()
+)
+
+def unit_param(*args, **kwargs):
+    return pytest.param(*args, **kwargs, marks=pytest.mark.unit)
+
+
+def quality_param(*args, **kwargs):
+    return pytest.param(*args, **kwargs, marks=pytest.mark.quality)
+
+
+def stress_param(*args, **kwargs):
+    return pytest.param(*args, **kwargs, marks=pytest.mark.stress)
+
+def make_cudf_series(arr):
+    df = pd.DataFrame(
+                           {'fea0': arr[:, ]})
+    df = df.values
+    df = df[:, 0]
+    return cudf.Series(df)
+
+def pickle_save_load(serializer_type, model):
+    os.mkdir('tmp')
+
+    if serializer_type == 'pickle':
+        serializer = pickle
+    else:
+        serializer = dill
+
+    pickle_file = 'tmp/cu_model'
+    try:
+        with open(pickle_file, 'wb') as pf:
+            serializer.dump(model, pf)
+    except (TypeError, ValueError) as e:
+        pf.close()
+        os.remove(pickle_file)
+        os.rmdir('tmp')
+        pytest.fail(e)
+
+    with open(pickle_file, 'rb') as pf:
+        cu_after_pickle_model = serializer.load(pf)
+
+    os.remove(pickle_file)
+    os.rmdir('tmp')
+
+    return cu_after_pickle_model
+
+@pytest.mark.parametrize('datatype', [np.float32])
+@pytest.mark.parametrize('input_type', ['dataframe'])
+@pytest.mark.parametrize('model', ['LinearRegression', 'Lasso', 'Ridge', 'ElasticNet'])
+@pytest.mark.parametrize('model_objects', [regression_models])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+@pytest.mark.parametrize('n_info', [unit_param(2)])
+@pytest.mark.parametrize('serializer_type', ['pickle', 'dill'])
+def test_regressor_pickle(datatype, input_type, model, model_objects, nrows, ncols, n_info,
+                          serializer_type):
+    train_rows = np.int32(nrows*0.8)
+    X, y = make_regression(n_samples=(nrows), n_features=ncols,
+                           n_informative=n_info, random_state=0)
+    X_test = np.asarray(X[train_rows:, :]).astype(datatype)
+    X_train = np.asarray(X[:train_rows, :]).astype(datatype)
+    y_train = np.asarray(y[:train_rows, ]).astype(datatype)
+
+    cu_before_pickle_model = model_objects[model]
+
+    if input_type == 'dataframe':
+        X_cudf = np_to_cudf(X_train)

```

Comment:
Could we just call these X_train, X_test, etc. like the non-cudf versions? Seems like the should be substitutable and then we can share the fit and predict lines without duplicating them on both branches of the `if`